### ISSUE TITLE
Support additional toolchain channels in rust-toolchain.toml? 

### ISSUE THREAD

**Issue by jmjoy** [rust-lang/rustup#3355]
### Problem you are trying to solve

Somethings we have to use multi toolchains in a project, usually stable and nightly, like:

```shell
cargo build --release
cargo +nightly fmt --all
cargo +nightly rustdoc 
```

Because we prefer to stable toolchain, but there are some useful features in nightly for `format` and `rustdoc`.

But, we can only use latest nightly version with `+nightly`, or hard code the nightly version `+nightly-2023-01-01`.

### Solution you'd like

Add the new field in `rust-toolchain.toml`, such as:

```toml
[toolchain.custom.foo]
channel = "nightly-2020-07-10"
components = [ "rustfmt", "rustc-dev" ]
targets = [ "wasm32-unknown-unknown", "thumbv2-none-eabi" ]
profile = "minimal"
```

and then, we can use this custom toolchain by:

```shell
cargo +foo fmt --all
```

### Notes

Maybe there's a better way.

**Comment by rbtcollins** [comment]
The challenge would be how to handle one of your additional aliases overlapping a custom toolchain. So for instance, a linked stage2 Rust toolchain with the alias `foo`.

Imagine you have a RUSTC_WRAPPER defined, which consults the `RUSTUP_TOOLCHAIN` to do something special based on the use of foo - it looks for RUSTUP_TOOLCHAIN=foo and does whatever special handling is appropriate.

Now, when I clone your repo that also has the alias `foo` to `stable` or whatever. 

In your working dir, we can handle resolving `foo` to `stable`. This works by setting `RUSTUP_TOOLCHAIN` to `stable` when the `rustup` proxies invoke the underlying commands like cargo. Now the RUSTC_WRAPPER doesn't see `foo` *at all*, it sees `stable`.

Thats fine right? But if you combine that with other automation, the reverse can happen.
In that case, commands would be run in your crate where the custom toolchain `foo` would show up.

My concern is that the user model folk work with is probably not this nuanced, and then there will be frustation and complexity.

That said, code wise this probably isn't too hard - we need to extend the resolution logic (if you go looking, look in master, we just overhauled it) - to treat a custom toolchain name as possibly indexing into a toolchain file, rather than short-circuiting that entirely.

**Comment by jmjoy** [comment]
It's really hard to deal with, I was sloppy.

I think the `rusutp toolchain link` has a similar effect, but the link is currently global and cannot target the current project directory.

First, if the link can support the project directory, for example, add the `.rustup` directory to the project directory, then `.rustup/toolchiains/foo` reference to another toolchain (this requires some modifications to the toolchian link).

Secondly, supports configure `[toolchain.link]` in `rust-toolchain.yaml`.

Depending on whether it is necessary to introduce a new directory `.rustup`.

**Comment by rbtcollins** [comment]
I don't think either of those would address the user confusion I expect from the asymmetricality of the situation. 

When a toolchain is refined down to a name like 'stable-x86_64-windows-msvc` information is lost - this is because rustup doesn't support aliases for toolchains : historically we had just one source of toolchains : ~/.rustup/toolchains. Then toolchain files were added, and they didn't change this because they could only name existing or installable toolchains. Then path based toolchains were added, and they also didn't change this because a path is never confusable with a fully qualified installable toolchain, and we added logic to prevent confusion with custom names.

I think for your use case I would recommend two routes.

Easiest: just use a pinned nightly in a toolchain file. You will have less recompilation except when picking a newer nightly to pin too, and you won't need to worry about compatibility with multiple toolchain versions, features only present in nightly etc.

Second, less easy: put a pinned nightly in a toolchain file, and for things you explicitly don't want to use nightly for, use +stable e.g. `cargo +stable doc`.
